### DISCUSSION TITLE
How to keep dynamically created options "alive" between form updates?

### DISCUSSION THREAD

**Discussion by henninghorn** [filamentphp/filament#8191] [Help]
Hi.

I'm building a form with two sets of a date picker input and a radio input.
One for departure and one for return. Here's a snippet:

```php
//...
DatePicker::make('departure_date')
    ->placeholder('Select date')
    ->closeOnDateSelection()
    ->minDate(now()->startOfDay())
    ->native(false)
    ->required()
    ->live(),
Radio::make('departure_flight')
    ->options(function (Get $get) {
        return $this->searchFlights($get('departure_date'));
    })
    ->hidden(fn (Get $get): bool => !$get('departure_date'))
//...
```

When I select a departure date, the departure_flight options gets dynamically updated as expected.
The radio input calls the searchFlights method and displays the returned options.

However, when I select a return date, both radio options callbacks gets executed.

How do I prevent this behavior, if possible? A simple solution could be to let the callback happen but also cache the searchFlights result, but I would rather only have the radio field to update its options when a specific other field gets updated.

I have also tried a different approach by adding a afterStateUpdated on the date picker field like so:
```php
//'''
->afterStateUpdated(function (?string $old, ?string $state, \Filament\Forms\Components\Component $component) {
    $component
        ->getContainer()
        ->getComponent('departure_flight', true)
        ->options([
            'a' => 'aaa: ' . $state,
            'b' => 'bbb: ' . $state,
        ]);
}),
//...
```

This successfully updates the radio options, hower the options are gone if a live() field is updated, or if you proceed to next step in a wizard.

Do anyone have a suggestion on how to keep dynamically generated options alive between form updates?

**Comment by danharrin** [comment]
The function passed to `options()` can `$get()` the value of the date picker and return the options. You should not set the options from `afterStateUpdated()`. https://filamentphp.com/docs/3.x/forms/advanced#dependant-select-options

**Reply by henninghorn** [reply]
I have no trouble getting the data from date picker field.

I have created an example form to illustrate the issue.

When you trigger a form update, the options will be re-generated using the function called within `options()`.

See it online here: [https://filament-v3-options-question.henninghorn.dk/](https://filament-v3-options-question.henninghorn.dk/)
See the form setup here: [https://github.com/henninghorn/filament-v3-options-question/blob/main/app/Livewire/OptionsExample.php](https://github.com/henninghorn/filament-v3-options-question/blob/main/app/Livewire/OptionsExample.php)

Can you see what I'm referring to? I want to prevent the function within `options()` to be called whenever the form updates.

**Reply by danharrin** [reply]
And I'm saying that you shouldn't prevent it from happening. You should not use afterStateUpdated() to set options. `options()` should always return the currently available options, whenever it is ran - hence why you need to fetch the date and calculate the options inside there. Filament handles the rest.

**Reply by henninghorn** [reply]
Okay, I'm not using the afterStateUpdated() anymore.

In the linked example I'm using `options()`:
```php
Radio::make('some_options')
    ->label("👆 These options are based on date above")
    ->options(function (Get $get) {
        return $this->generateOptionsFromApi($get('some_date'));
    })
    ->hidden(fn (Get $get): bool => !$get('some_date')),
```

As I understand it, Filament will always re-run the function in a fields' `options()`, correct?
And if I by some reason do not want to call an external API when an unrelated field gets updated, should I then have som caching  mechanism in order to prevent an unnecessary external API request?

I just want to make sure to do it the right way :)

**Reply by danharrin** [reply]
Yeah, a caching mechanism sounds good for this

**Reply by henninghorn** [reply]
Cool, thank you for your input! Filament is such a joy to work with 🙌
